syntax = "proto3";

package sigprocpacket;

message SigProcPacket {
    repeated RXChan rxchannel = 1; // REVIEW #28 spec?
    repeated DMStage decimation_stages = 2;
    uint32 sequence_num = 3;
    double kerneltime = 4;
    float sequence_time = 5;
    uint32 offset_to_first_rx_sample = 6;
    double rxrate = 7;
    double output_sample_rate = 8;

    message RXChan {
        uint32 slice_id = 1;
        double rxfreq = 2; // this is offset frequency from centre freq, kHz
        repeated BeamDirections beam_directions = 3;
        uint32 num_ranges = 4;
        float first_range = 5; // km
        float range_sep = 6; // km
        uint32 tau_spacing = 7; // in us
        repeated LagPair lags = 8; // will be empty if no acfs.


        message BeamDirections {
            repeated Phase phase = 1; /* must be length = number main + intf antennas */

            message Phase {
                double real_phase = 1;
                double imag_phase = 2;
            }
        }

        message LagPair {
            uint32 pulse_1 = 1;
            uint32 pulse_2 = 2;
            uint32 lag_num = 3;
        }
    }

    message DMStage {
        uint32 stage_num = 1;
        double input_rate = 2;
        uint32 dm_rate = 3;
        repeated float filter_taps = 4;
    }

}
